ray-curator/ray_curator/stages/deduplication/fuzzy/connected_components.py

```diff
@@ -0,0 +1,209 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from typing import TYPE_CHECKING, Any
+
+import cudf
+import cugraph
+from cugraph.dask.comms.comms_wrapper import init_subcomms as c_init_subcomms
+from loguru import logger
+from pylibcugraph import GraphProperties, MGGraph, ResourceHandle
+from pylibcugraph import weakly_connected_components as pylibcugraph_wcc
+
+from ray_curator.backends.experimental.utils import RayStageSpecKeys
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.deduplication.id_generator import (
+    CURATOR_DEDUP_ID_STR,
+)
+from ray_curator.stages.deduplication.io_utils import DeduplicationIO
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks.file_group import FileGroupTask
+from ray_curator.utils.file_utils import delete_dir, get_fs, is_not_empty
+
+if TYPE_CHECKING:
+    from ray_curator.backends.base import WorkerMetadata
+
+
+class ConnectedComponentsStage(ProcessingStage[FileGroupTask, FileGroupTask], DeduplicationIO):
+    def __init__(
+        self,
+        output_dir: str,
+        source_column: str = f"{CURATOR_DEDUP_ID_STR}_x",
+        destination_column: str = f"{CURATOR_DEDUP_ID_STR}_y",
+        read_kwargs: dict | None = None,
+        write_kwargs: dict | None = None,
+    ):
+        """
+        Args:
+            output_dir: The directory to write the resulting connected components to.
+            source_column: The column name containing the document ids of the source of the edge.
+            destination_column: The column name containing the document ids of the destination of the edge.
+            read_kwargs: Keyword arguments to pass for reading the input files.
+            write_kwargs: Keyword arguments to pass for writing the output files.
+        """
+        # Initialize parent classes
+        ProcessingStage.__init__(self)
+        DeduplicationIO.__init__(self, id_generator=None)
+
+        self.source_column = source_column
+        self.destination_column = destination_column
+        self.read_kwargs = read_kwargs if read_kwargs is not None else {}
+        self.write_kwargs = write_kwargs if write_kwargs is not None else {}
+
+        self._name = self.__class__.__name__
+        self._resources = Resources(cpus=1.0, gpus=1.0)
+        self._batch_size = None
+
+        # Handle output directory cleanup logic
+        self.output_fs = get_fs(output_dir, self.write_kwargs.get("storage_options", {}))
+        self.output_dir = self.output_fs.sep.join([output_dir, self.name])
+        if is_not_empty(self.output_dir, self.output_fs):
+            logger.warning(f"Output directory {self.output_dir} is not empty. Deleting it.")
+            delete_dir(self.output_dir, self.output_fs)
+        self.output_fs.mkdirs(self.output_dir, exist_ok=True)
+
```

Comment:
Can you move this to a file_utils? I could also use it in kmeans / pairwise

Reply:
We had something like this here

https://github.com/ayushdg/ray-curator/blob/92d1280010e9953ef8c602ea9441aaf9fecfddeb/ray_curator/utils/file_utils.py#L63C5-L76

FWIW s3's mkdirs doesn't work IIRC, which is why the code above has `fs.touch(os.path.join(dir_path, ".empty"))`

Reply:
Based on the comments here: https://github.com/fsspec/s3fs/issues/401 mkdirs is a no-op for s3 which is fine since directly writing out to a path without making directories is allowed. I don't think a `touch` is needed in this case.